src/sql/workbench/services/objectExplorer/browser/objectExplorerService.ts

```diff
@@ -407,67 +408,92 @@ export class ObjectExplorerService implements IObjectExplorerService {
-								// For folders send the actual name of the folder (since the nodeTypeId isn't useful in this case and the names are controlled by us)
-								const nodeType = node.nodeTypeId === NodeType.Folder ? node.label : node.nodeTypeId;
-								this._telemetryService.createErrorEvent(TelemetryKeys.TelemetryView.Shell, TelemetryKeys.TelemetryError.ObjectExplorerExpandError, undefined, errorType)
-									.withAdditionalProperties({
-										nodeType,
-										providerId: expandResult.providerId
-									}).send();
+					try {
+						let resultMap: Map<string, azdata.ObjectExplorerExpandInfo> = new Map<string, azdata.ObjectExplorerExpandInfo>();
+						let allProviders: azdata.ObjectExplorerProviderBase[] = [provider];
+
+						let nodeProviders = this._nodeProviders[providerId];
+						if (nodeProviders) {
+							nodeProviders = nodeProviders.sort((a, b) => a.group!.toLowerCase().localeCompare(b.group!.toLowerCase()));
+							allProviders.push(...nodeProviders);
+						}
+						this.logService.trace(`${session.sessionId}: got providers for node expansion: ${allProviders.map(p => p.providerId).join(', ')}`);
+
+						let retryCount = 0;
+						const expansionTimeout = this._configurationService.getValue<number>('serverTree.nodeExpansionTimeout');
+						const timeout = setTimeout(() => {
+							/**
+							 * If we don't get a response back from the provider in specified expansion timeout seconds then we assume
+							 * it's not going to respond and resolve the promise with the results we have so far
+							 */
+							if (resultMap.size === allProviders.length || retryCount === expansionTimeout) {
+								resolve(self.mergeResults(allProviders, resultMap, node.nodePath));
+								if (newRequest) {
+									delete self._sessions[session.sessionId!].nodes[node.nodePath];
+									this.logService.trace(`Deleted node ${node.nodePath} from session ${session.sessionId}`);
+								}
+								clearTimeout(timeout);
 							}
+							retryCount++;
+						}, 1000);
+
+						self._sessions[session.sessionId!].nodes[node.nodePath].expandEmitter.event((expandResult: NodeExpandInfoWithProviderId) => {
+							if (expandResult && expandResult.providerId) {
+								this.logService.trace(`${session.sessionId}: Received expand result for node ${node.nodePath} from provider ${expandResult.providerId}`);
+								resultMap.set(expandResult.providerId, expandResult);
+								// If we got an error result back then send error our error event
+								// We only do this for the MSSQL provider
+								if (expandResult.errorMessage) {
+									const errorType = expandResult.errorMessage.indexOf('Object Explorer task didn\'t complete') !== -1 ? 'Timeout' : 'Other';
+									// For folders send the actual name of the folder (since the nodeTypeId isn't useful in this case and the names are controlled by us)
+									const nodeType = node.nodeTypeId === NodeType.Folder ? node.label : node.nodeTypeId;
```

Comment:
We don't know what other providers will send so let's just only do the folder name for MSSQL, we can add others later on on a case-by-case basis if needed

```suggestion
									// For MSSQL folders send the actual name of the folder (since the nodeTypeId isn't useful in this case and the names are controlled by us)
									const nodeType = expandResult.providerId === mssqlProviderName && node.nodeTypeId === NodeType.Folder ? node.label : node.nodeTypeId;
```

Reply:
(you'll also need to add back in the import, so this might be easier to do manually yourself and then push as a single commit)